### PR TITLE
query-frontend: cancel stream to scheduler when streaming failed

### PR DESCRIPTION
**Pull request by dimitarvdimitrov**
Signed-off-by: Dimitar Dimitrov <dimitar.dimitrov@grafana.com>

#### What this PR does

discovered by @pstibrany:

makes sure that the query-frontend properly terminates gRPC streams to the query-scheduler after the scheduler closes the connections on them

without this there was a goroutine and memory leak. In the attached screenshot you can see on the left side of each graph GEM running without this fix and on the right side of each graph GEM with this fix. 
Screenshot 2022-10-25 at 10 37 01

#### Checklist

- [ ] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Review by pstibrany — COMMENTED** [review]
Looks good. I have one comment, and a proposal for a test:

https://gist.github.com/pstibrany/7a4384aee0b8389475ca7dd34af58410#file-frontend_test-go-L437

**Comment by dimitarvdimitrov** [comment]
thanks for the review and the test @pstibrany. I think I've incorporated the changes you requested and the test too. Can you take another look?

File: pkg/frontend/v2/frontend_test.go [hunk 1]
```diff
@@ -125,6 +132,7 @@ func TestFrontendBasicWorkflow(t *testing.T) {
 	require.NoError(t, err)
 	require.Equal(t, int32(200), resp.Code)
 	require.Equal(t, []byte(body), resp.Body)
+	t.Cleanup(f.StopAsync)
```
Comment (pstibrany) [inline pkg/frontend/v2/frontend_test.go:135]: Why do we call `f.StopAsync` here, but also `services.StopAndAwaitTerminated(context.Background(), f)` (which stops the same service and waits until it has stopped)?

(I'd keep previous code, calling `_ = services.StopAndAwaitTerminated(context.Background(), f)` in the Cleanup function, and remove this line.)

File: pkg/frontend/v2/frontend_scheduler_worker.go [hunk 2]
```diff
@@ -266,18 +266,15 @@ func (w *frontendSchedulerWorker) stop() {
 }
 
 func (w *frontendSchedulerWorker) runOne(ctx context.Context, client schedulerpb.SchedulerForFrontendClient) {
-	backoffConfig := backoff.Config{
-		MinBackoff: 250 * time.Millisecond,
-		MaxBackoff: 2 * time.Second,
-	}
+	// attemptLoop returns false iff it couldn't properly start or end the loop
```
Comment (pstibrany) [inline pkg/frontend/v2/frontend_scheduler_worker.go:269]: Not sure if this comment is correct: "end the loop"... what about:

```suggestion
	// attemptLoop returns false if there was any error with forwarding requests to scheduler.
```

**Review by pstibrany — APPROVED** [review]
Looks good, thank you!

File: pkg/frontend/v2/frontend_test.go (hunk 1 above)
Comment (dimitarvdimitrov) [reply pkg/frontend/v2/frontend_test.go:135]: oh, I committed this by mistake, sorry

File: pkg/frontend/v2/frontend_test.go [hunk 3]
```diff
@@ -425,3 +432,62 @@ func TestConfig_Validate(t *testing.T) {
...
+	resp, err = f.RoundTripGRPC(user.InjectOrgID(context.Background(), userID), &httpgrpc.HTTPRequest{})
+	require.NoError(t, err)
+	require.Equal(t, int(resp.Code), http.StatusTooManyRequests)
+
+	// There should still be only one stream open, and one goroutine created for it.
+	// Previously frontend leaked goroutine because stream that received "EOF" due to server closing the connection, never stopped its goroutine.
+	require.Equal(t, 1, checkStreamGoroutines())
+}
+
+const createdBy = "created by google.golang.org/grpc.newClientStreamWithParams"
```
Comment (pracucci) [inline pkg/frontend/v2/frontend_test.go:475]: [nit] Move this const inside `checkStreamGoroutines()`.

**Review by pracucci — APPROVED** [review]
Great job! I just let a nit 🙏

**Pull request merged**